lib/class-wp-webfonts.php

```diff
@@ -0,0 +1,323 @@
+			self::$webfonts[ $id ] = $font;
+		}
+	}
+
+	/**
+	 * Get the font ID.
+	 *
+	 * @param array $font The font arguments.
+	 * @return string
+	 */
+	public function get_font_id( $font ) {
+		return sanitize_title( "{$font['font-family']}-{$font['font-weight']}-{$font['font-style']}-{$font['provider']}" );
+	}
+
+	/**
+	 * Validate a font.
+	 *
+	 * @param array $font The font arguments.
+	 *
+	 * @return array|false The validated font arguments, or false if the font is invalid.
+	 */
+	public function validate_font( $font ) {
+		$font = wp_parse_args(
+			$font,
+			array(
+				'provider'     => 'local',
+				'font-family'  => '',
+				'font-style'   => 'normal',
+				'font-weight'  => '400',
+				'font-display' => 'fallback',
+			)
+		);
+
+		// Check the font-family.
+		if ( empty( $font['font-family'] ) || ! is_string( $font['font-family'] ) ) {
+			trigger_error( __( 'Webfont font family must be a non-empty string.', 'gutenberg' ) );
+			return false;
+		}
+
+		// Local fonts need a "src".
+		if ( 'local' === $font['provider'] ) {
+			// Make sure that local fonts have 'src' defined.
+			if ( empty( $font['src'] ) || ( ! is_string( $font['src'] ) && ! is_array( $font['src'] ) ) ) {
+				trigger_error( __( 'Webfont src must be a non-empty string or an array of strings.', 'gutenberg' ) );
+				return false;
+			}
+		}
+
+		// Validate the 'src' property.
+		if ( ! empty( $font['src'] ) ) {
+			foreach ( (array) $font['src'] as $src ) {
+				if ( empty( $src ) || ! is_string( $src ) ) {
+					trigger_error( __( 'Each webfont src must be a non-empty string.', 'gutenberg' ) );
+					return false;
+				}
+
+				if (
+					// Validate data URLs.
+					! preg_match( '/^data:.+;base64/', $src ) &&
+					// Validate URLs.
+					! filter_var( $src, FILTER_VALIDATE_URL ) &&
+					// Check if it's a URL starting with "//" (omitted protocol).
+					0 !== strpos( $src, '//' )
+				) {
+					trigger_error( __( 'Webfont src must be a valid URL or a data URI.', 'gutenberg' ) );
+					return false;
+				}
+			}
+		}
+
+		// Check the font-style.
+		$valid_font_styles = array( 'normal', 'italic', 'oblique', 'inherit', 'initial', 'revert', 'unset' );
+		if ( ! in_array( $font['font-style'], $valid_font_styles, true ) && ! preg_match( '/^oblique\s+(\d+)%/', $font['font-style'] ) ) {
+			return false;
+		}
+
+		// Check the font-weight.
+		if ( // Bail out if the font-weight is not a valid value.
+			( ! is_string( $font['font-weight'] ) && ! is_int( $font['font-weight'] ) ) ||
+			(
+				// Check if value is a single font-weight, formatted as a number.
+				! in_array( $font['font-weight'], array( 'normal', 'bold', 'bolder', 'lighter', 'inherit' ), true ) &&
+				// Check if value is a single font-weight, formatted as a number.
+				! preg_match( '/^(\d+)$/', $font['font-weight'], $matches ) &&
+				// Check if value is a range of font-weights, formatted as a number range.
+				! preg_match( '/^(\d+)\s+(\d+)$/', $font['font-weight'], $matches )
+			)
```

Comment:
Same as above, not sure this code is needed and the behaviour it implements is not the best imho (PHP error log vs. browser console errors).

Reply:
fixed 👍